### PR TITLE
MariaDB - Galera implementation integrated with KPM

### PR DESCRIPTION
**Pull request by mateusz-blaszkowski**
Signed-off-by: Mateusz Blaszkowski <mateusz.blaszkowski@intel.com>

### CODE REVIEW

File: mariadb-galera/manifest.jsonnet [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+function(
+  params={}
+)
+
+kpm.package({
+  package: {
+    name: "stackanetes/mariadb-galera",
+    expander: "jinja2",
+    author: "Quentin Machu",
```
Comment (DTadrzak) [inline mariadb-galera/manifest.jsonnet]: I'm not sure whether Quentin is Author of this package.;]

File: mariadb-galera/templates/daemonset.yaml.j2 [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        - name: pid
+          configMap:
+            name: mariadb-galera-pid
+        - name: tuning
+          configMap:
+            name: mariadb-galera-tuning
+        - name: wsrep
+          configMap:
+            name: mariadb-galera-wsrep
+        - name: mysql
```
Comment (DTadrzak) [inline mariadb-galera/templates/daemonset.yaml.j2]: We shouldn't mount it to emptyDir.

File: mariadb-galera/templates/init.yaml.j2 [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        - name: pid
+          configMap:
+            name: mariadb-galera-pid
+        - name: tuning
+          configMap:
+            name: mariadb-galera-tuning
+        - name: wsrep
+          configMap:
+            name: mariadb-galera-wsrep
+        - name: mysql
```
Comment (DTadrzak) [inline mariadb-galera/templates/init.yaml.j2]: I'm not sure whether we need to mount mysql to init mysql job isntance.

File: mariadb/manifest.jsonnet [hunk 4]
```diff
@@ -51,25 +51,96 @@ kpm.package({
   resources: [
     // Config maps.
     {
-      file: "configmaps/my.cnf.yaml.j2",
-      template: (importstr "templates/configmaps/my.cnf.yaml.j2"),
-      name: "mariadb-mycnf",
+      file: "configmaps/seed.sh.yaml.j2",
+      template: (importstr "templates/configmaps/seed.sh.yaml.j2"),
+      name: "seedsh",
```
Comment (Quentin-M) [inline mariadb/manifest.jsonnet]: Could we keep the `mariadb` prefix? We already have over 50 configmaps: it would get messy/confusing without it and conflicts might happen (especially for the `startsh` below!).

File: mariadb/manifest.jsonnet [hunk 5]
```diff
@@ -51,25 +51,96 @@ kpm.package({
...
+      name: "mariadb-init",
+      type: "job",
+    },
+
+    // DaemonSets.
+    {
+      file: "daemonset.yaml.j2",
+      template: (importstr "templates/daemonset.yaml.j2"),
+      name: "mariadb-ds",
+      type: "daemonset",
```
Comment (Quentin-M) [inline mariadb/manifest.jsonnet]: Do we want to make it a deployment with a set number of replicas (e.g. quorum of 3, 5) or really a daemonset?

File: mariadb/manifest.jsonnet [hunk 6]
```diff
@@ -51,25 +51,96 @@ kpm.package({
...
+      template: (importstr "templates/init.yaml.j2"),
+      name: "mariadb-init",
+      type: "job",
+    },
+
+    // DaemonSets.
+    {
+      file: "daemonset.yaml.j2",
+      template: (importstr "templates/daemonset.yaml.j2"),
+      name: "mariadb-ds",
```
Comment (Quentin-M) [inline mariadb/manifest.jsonnet]: `mariadb` is enough.

File: mariadb/templates/configmaps/seed.sh.yaml.j2 [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+kind: ConfigMap
+data:
+  seed.sh: |+
+    #!/bin/sh
+
+    set -ex
+
+    SLEEP_TIMEOUT=5
+    # we need to count seed instance here.
+    MINIMUM_CLUSTER_SIZE=4
```
Comment (Quentin-M) [inline mariadb/templates/configmaps/seed.sh.yaml.j2]: Should it be in relation with the replicas variable defined in stackanetes instead of being hardcoded?

Comment (mateusz-blaszkowski) [inline mariadb/templates/configmaps/seed.sh.yaml.j2]: Minimum sensible cluster size is always 3 (4 in this case cause the seed instance is also included). When 3-instance cluster is built, then there's no need to keep seed instance running.

File: mariadb/manifest.jsonnet (hunk 5 above)
Comment (mateusz-blaszkowski) [inline mariadb/manifest.jsonnet]: Agreed with the team (@ss7pro) to leave it as a DaemonSet for now. It'd require some additional work to setup Kubernetes 1.4 (for MariaDB running as a Deployment it's necessary to have anti-affinity feature which prevents the PODs to be scheduled on the same node). Therefore I'll create a new Jira item to replace current DaemonSet implementation with Deployment - it'll depend on Kubernetes 1.4 setup item.

File: mariadb/templates/configmaps/seed.sh.yaml.j2 (hunk 7 above)
Comment (Quentin-M) [inline mariadb/templates/configmaps/seed.sh.yaml.j2]: Let me rephrase: what happens if an user wants to simply try Stackanetes out with 2 nodes (1 control, 1 compute, `deployment.replicas=1`)? Can we do it?

File: mariadb/templates/configmaps/start.sh.yaml.j2 [hunk 8]
```diff
@@ -4,34 +4,9 @@ data:
   start.sh: |+
     #!/bin/bash
     set -ex
+    trap "trap - SIGTERM && kill -- -$$" SIGINT SIGTERM EXIT
 
-    # Setup folders and permissions.
-    chown mysql: /var/lib/mysql
+    chown mysql: /var/lib/mysql/openstack
```
Comment (ss7pro) [inline mariadb/templates/configmaps/start.sh.yaml.j2]: I would rename this directory to mysql-openstack-{{{ database.cluster_name }} so we would allow for multiple mariadb deployments on the same set of nodes.

File: mariadb/templates/init.yaml.j2 [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+kind: Job
+spec:
+  template:
+    metadata:
+      labels:
+        app: mariadb
+    spec:
+      restartPolicy: Never
+      # TODO(DTadrzak): it must be removed in future
+      securityContext:
```
Comment (ss7pro) [inline mariadb/templates/init.yaml.j2]: pls drop it.

File: mariadb/templates/init.yaml.j2 [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+        runAsUser: 0
+      terminationGracePeriodSeconds: 10000
+      containers:
+        - name: mariadb-init
+          image: {{ deployment.image.mariadb }}
+          imagePullPolicy: Always
+          # TODO(DTadrzak): it must be removed in future
+          securityContext:
```
Comment (ss7pro) [inline mariadb/templates/init.yaml.j2]: this should be dropped as well.

File: mariadb/templates/init.yaml.j2 [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+              valueFrom:
+                fieldRef:
+                  fieldPath: metadata.name
+            - name: NAMESPACE
+              valueFrom:
+                fieldRef:
+                  fieldPath: metadata.namespace
+            - name: COMMAND
+              value: "bash /tmp/seed.sh"
+            - name: ANSIBLE_LIBRARY
```
Comment (ss7pro) [inline mariadb/templates/init.yaml.j2]: this could be dropped as well.

File: mariadb/templates/daemonset.yaml.j2 [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+  template:
+    metadata:
+      labels:
+        app: mariadb
+        galera: enabled
+    spec:
+      nodeSelector:
+        {{ deployment.control_node_label }}: enabled
+      # TODO(DTadrzak): it must be removed in the future
+      securityContext:
```
Comment (ss7pro) [inline mariadb/templates/daemonset.yaml.j2]: remove it pls.

File: mariadb/templates/configmaps/seed.sh.yaml.j2 (hunk 7 above)
Comment (Quentin-M) [inline mariadb/templates/configmaps/seed.sh.yaml.j2]: It do believe that it is very important to keep support for minimal deployments. Stackanetes is really well adapted for developers today and it should remain that way. It is also useful to do workshops/demos/POC, where people only have time for 2 nodes deployment at most.

File: mariadb/templates/daemonset.yaml.j2 (hunk 12 above)
Comment (mateusz-blaszkowski) [inline mariadb/templates/daemonset.yaml.j2]: MariaDB needs permissions to write to data directory (`/var/lib/mysql-openstack-{{ database.cluster_name }}`). This directory is created as a `hostPath` type of volume - which is always mounted as root. According to https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/volumes.md: 

> The hostPath volume should only be used by effective-root users, and the permissions of paths exposed into containers via hostPath volumes should always be managed by the cluster operator.

File: mariadb/manifest.jsonnet [hunk 13]
```diff
@@ -51,9 +51,9 @@ kpm.package({
   resources: [
     // Config maps.
     {
-      file: "configmaps/my.cnf.yaml.j2",
-      template: (importstr "templates/configmaps/my.cnf.yaml.j2"),
-      name: "mariadb-mycnf",
+      file: "configmaps/seed.sh.yaml.j2",
+      template: (importstr "templates/configmaps/seed.sh.yaml.j2"),
+      name: "mariadb-seedsh",
       type: "configmap",
```
Comment (DTadrzak) [inline mariadb/manifest.jsonnet]: fix that please !

File: mariadb/manifest.jsonnet [hunk 14]
```diff
@@ -78,7 +149,7 @@ kpm.package({
       template: (importstr "templates/service.yaml.j2"),
       name: "mariadb",
       type: "service",
-    },
+    }
```
Comment (DTadrzak) [inline mariadb/manifest.jsonnet]: not needed !

File: mariadb/templates/configmaps/pid.cnf.yaml.j2 [hunk 15]
```diff
@@ -0,0 +1,6 @@
+apiVersion: v1
+kind: ConfigMap
+data:
+  pid.cnf: |+
```
Comment (DTadrzak) [inline mariadb/templates/configmaps/pid.cnf.yaml.j2]: LGTM

File: mariadb/templates/daemonset.yaml.j2 [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+        runAsUser: 0
+      containers:
+        - name: mariadb
+          image: {{ deployment.image.mariadb }}
+          imagePullPolicy: Always
```
Comment (DTadrzak) [inline mariadb/templates/daemonset.yaml.j2]: dont use Always !

File: mariadb/templates/service.yaml.j2 [hunk 17]
```diff
@@ -2,6 +2,6 @@ apiVersion: v1
 kind: Service
```
Comment (DTadrzak) [inline mariadb/templates/service.yaml.j2]: Kind: Pod

**Comment by PiotrProkop** [comment]
Can you add mariadb-init job as a dependency to services using mariadb  ?

**Comment by PiotrProkop** [comment]
@mateusz-blaszkowski please resolve confilcts.

**Comment by PiotrProkop** [comment]
Look's good.Merging.
